README.md
# FF.PyAdmin (接口服务/后台管理微框架)

## 初衷

运维开发中, 各类接口服务和后台管理系统必不可少, 索性写个通用的/简单易用的/支持快速开发的微框架.

一是开发时尽可能只关注业务逻辑, 快速实现, 不用再过多考虑架构/前端/布局什么的.

二是将常用业务逻辑二次封装, 使用时, 用简洁的语句即可完成复杂的事务, 并自动处理异常.

三是方便运维小伙伴们快速理解 Web 项目开发, 无须深入理解 Flask 和前端即可完成此类项目开发.

## 特点

1. 后端 Flask, 对异常处理, 数据库, 表单验证, 权限验证都做了二次封装, 调用更简单, 通用性强.
2. 前端 LayUI, 对数据表格作了二次封装, 支持自动按数据表生成表头, 可选使用. 丰富配色, 丰富图标.
3. 前后端完全使用 API 接口通信, JSON 数据交互. 方便与各类前端框架/小程序对接.
4. 适合做纯 API 接口服务, 也适合同时做接口服务和后台管理, 应用广泛/灵活.
5. 对 HTTP 异常二次封装, APIException/MsgException, 标准化 JSON 字段和错误消息页面.
6. 实现全局异常处理, 支持配置中自定义 404 等状态码消息. 500 状态码时自动记录日志.
7. 异常返回自适应, POST 和 AJAX 请求自动返回JSON数据, 其他状态展示自定义错误页, DEBUG 显示原始异常.
8. 自定义常用权限校验装饰器, 登录验证, 菜单权限验证, 接口权限验证, 来访 IP 白名单等.
9. 依托 Flask 蓝图和视图函数, 适应大部分情况下的权限管理需求. 扩展性强.
10. WTForms 二次封装, 极简调用, 适应性强. 中文内置错误消息, 自动数据获取的验证.
11. 验证失败自动按需返回 JSON 或错误页. 自定义了通用的字符串和正整数验证方法, 适用性广.
12. SQLAlchemy 数据库交互二次封装, 实现自动提交机制和完善的异常处理, CRUD 全封装, 极简调用.
14. 自带文件日志, 按日志文件大小自动滚动, 无须配置.
15. 安全, 方便, 完善的配置文件方案.

## 图示

![./docs/FF.PyAdmin.Code.png](./docs/FF.PyAdmin.Code.png)

![./docs/FF.PyAdmin.png](./docs/FF.PyAdmin.png)

![./docs/update.v1.2.png](./docs/update.v1.2.png)

## 依赖

1. pycryptodome==3.9.8
2. Authlib==0.14.3
3. behave==1.2.6
4. blinker==1.4
5. concurrent-log-handler==0.9.16
6. cymysql==0.9.14
7. requests==2.24.0
8. SQLAlchemy==1.3.18
9. Werkzeug==1.0.1
10. WTForms==2.3.1
11. email-validator==1.1.1
12. xxtea==2.0.0.post0
13. Flask==1.1.2
14. Flask-Login==0.5.0
15. Flask-Mail==0.9.1
16. Flask-Migrate==2.5.3
17. Flask-Script==2.0.6
18. Flask-SQLAlchemy==2.4.3
19. Flask-WTF==0.14.3
20. PS: 升级了相关软件包到最新, 旧依赖包见 `requirements.old.txt` 同样兼容.

该框架基于公司内部 OA 系统 OAuth2 登录, 在登录画面点击即可进入演示后台. 需要自行准备 OAuth2 参数配置到配置文件, 如 QQ, 微信, GitHub, 钉钉等. 也可自行实现账号密码登录.

```python
# app/views/web.py
@bp_web.route('/authorize')
def web_authorize():
    """OAuth 登录跳转"""
    # TODO: (演示使用, 自动登录), 请删除并配置自己的认证方式, OAuth2或账密系统
    set_user_login({
        'job_number': 7777,
        'realname': 'Fufu'
    })
    return redirect(url_for('web.web_index'))
```



## 结构

    .
    ├── app  应用根目录, app.root_path
    │   ├── conf  配置文件目录
    │   ├── forms  表单验证
    │   ├── libs  公共类库
    │   ├── models  表模型
    │   ├── services  中间件, 逻辑处理
    │   ├── static  静态文件
    │   ├── templates  模板文件
    │   └── views  视图函数
    ├── data  数据文件
    ├── docs  开发文档
    ├── features  BDD
    │   └── steps
    ├── logs  日志目录
    │   └── behave.reports
    ├── scripts  脚本, 配置文件
    ├── tests  单元测试
    ├── tmp  临时文件
    └── venv  虚拟环境

## 配置

1. `secret_settings.py` 必须, 一般设置 SECRET_KEY 及数据库配置, 不要加入 Git.
2. `settings.py` 必须, 常规配置, APP_NAME, 日志参数, 错误消息等.
3. `production_settings.py` 可选, 默认加载项, 生产环境额外配置.
4. `development_settings.py` 可选, 开发环境额外配置, 不建议加入 Git.
5. `testing_settings.py` 可选, 测试环境额外配置, BDD 或单元测试时使用, 不建议加入 Git.

默认只会加载第 1, 2 项配置, 第 3/4/5 项需要设置环境变量, 可选: `production` `development` `testing`. 如:

```shell
# 1. Windows:
	set FF_PyAdmin=development
	echo %FF_PyAdmin%
# 2. Linux:
	export FF_PyAdmin=development
	echo $FF_PyAdmin
```

生产环境建议使用加密环境变量, 生成加密数据的方式如下:

```shell
# Fufu~~Text 加密后 792e01268cc350077f772a5f
python3 manage.py encrypt -d Fufu~~Text
```

使用方法见 `secret_settings.py` 中 `OA_CLIENT_SECRET` 和 `PYADMIN_DBPASS`, 配置示例:

```shell
# Linux (Windows 使用上面的 set 方法)
cp scripts/etc-profile.d-ffpyadmin.sh /etc/profile.d/ffpyadmin.sh
chmod +x /etc/profile.d/ffpyadmin.sh
source /etc/profile.d/ffpyadmin.sh
```

也可以使用不加密的环境变量, 配置文件中的获取环境变量时不加 key 参数即可:

```python
# OA_CLIENT_SECRET = get_environ('PYADMIN_OAUTH_SECRET', key=SECRET_KEY)
OA_CLIENT_SECRET = get_environ('PYADMIN_OAUTH_SECRET')
```

**建议开发环境直接使用快捷方式启动: `python3 dev.py`**

## 使用

1. Python3.5+(Python3.8测试通过), pip, MySQL 环境
2. 建好 MySQL 数据库, 导入 data/db_ff_pyadmin.sql 示例数据, 修改配置文件
3. `pip3 install -r requirements.txt`
4. `python3 start.py` 或 `python3 dev.py`
5. 访问: http://127.0.0.1:5000  [http://ff.pyadmin:777](http://ff.pyadmin:777/)
6. 生产环境部署方案见 `script`

## 资料

- Flask https://github.com/pallets/flask
- OAuth2 https://github.com/lepture/authlib
- LayUI https://github.com/sentsin/layui


CHANGELOG.md
# FF.PyAdmin

## v1.2.3

**2021-01-28 更新:**

- 修正自定义异常描述显示问题
- 新增访问第三方数据库的快捷方式, 数据源与逻辑分离, 详见: `/result_tpl`, 配置见: `development_settings.py`
  - `OCI` Oracle 需要下载客户端环境, 安装 `cx-oracle`
  - `MSSQL` SQLServer 推荐使用 `pyodbc`, 也可以用 `pymssql`
  - `MYSQL` 驱动任选
- 助手函数新增 `get_json_loads`, `get_iso_date`
- 助手函数 `get_date` 增加秒参数, 增加输出时间戳(毫秒), 增加输出时间日期对象

## v1.2.2

**2020-09-21 更新:**

- 助手函数 `get_date` 增加支持传入时间戳
- 增加 `DEBUG_RESPONSE` 配置参数, 开关是否在 DEBUG` 模式下启用响应日志
- API 返回结果增加 `info` 字段, 无论成功或失败都可以附带该信息
- 新版 WTForms 同样有例外问题, 启用自定义 `CSRFProtect`
- SQLAlchemy 增加默认参数: `max_overflow': -1` 
- 新增 `init_db_cache` 机制, 用于将更新频率低, 使用频率高的数据存入全局变量(如 IP 白名单, 示例中是 ASN 表: `/show_asn_cache`), 并实时更新, 应用启动后不再请求数据库
- 视图函数可直接返回集合数据类型 `return set()`
- 表单验证 `StripString()` 新增 `cls_whitespace=True` 用于清除所有空白字符
- 权限禁用 `role_deny` 限制只能禁用小权限(视图函数级别)


## v1.2.1

**2020-08-14 更新:**

- 增加助手函数 `get_real_ip`, 方便获取客户端 IP
- 增加部署方案示例配置: `script/gunicorn` `script/uwsgi`

## v1.2

**2020-07-10 更新:**

- 升级所有包为最新版本(Python3.8测试通过): `Flask==1.1.2, WTForms==2.3.1`, 旧依赖库(正常可用): `requirements.old.txt`
- 升级 `LayUI==2.5.6`
- 自动表格 `$.autoTbl` 增加属性 `colsFields`, 可指定要显示的表格字段. 增加属性 `colsAlign` 指定默认对齐方式
- 新增示例: 真分页, 真排序, 前端增加字段, 后端隐藏字段, 指定显示字段, 控制表格字段显示顺序, 指定对齐方式等, 详见 `/bgp` 及前后端注释
- 异步请求 `$.mkAjax` 增加回调方法 `done_err`, Ajax 请求成功但返回 `ok!=1` 时可触发
- 增加信号中异步任务示例: `/async_signal_demo`
- 使用 `flask_wtf` 原生 `CSRFProtect`, 增加例外示例: `/log/csrf_exempt_demo`
- 表单校验增加 `FloatRound` 用于校验浮点数, 金额等, 银行家舍入法, 默认保留2位小数
- 增加默认消息页面: `base-msg.html`
- 增加开发环境快捷启动方式: `python3 dev.py`, `PyCharm Script path` 可指向该文件
- 增加助手函数: `get_round` `get_ymd` `get_next_month_first` `get_month_last` `get_last_month_last`
- 增加模板函数: `get_date`, 示例表单默认昨天日期: `<input value="{{ get_date(out_fmt="%Y-%m-%d", add_days=-1) }}">`
- 对称加密类库更新: `AESCipher`, 增加助手函数, 用法详见注释
- `flask.json` 增加 `JSONEncoder` 对 `Decimal` 支持
- 数据库日志服务改用 `flask.json`, 以适应常见各类数据类型
- ORM 增加 `hide_keys_dicts` 方法, 将 `.all()` 转为字典并隐藏指定字段
- 调试模式且 Ajax POST 请求时将 Response JSON 显示在控制器方便调试
- 修正 `logger` 文件日志级别

## v1.1

**2019-12-20 更新:**

- 新增配置项: `MAIL_OPEN` 邮件发送开关, 与邮件配置项配合使用, 默认 `False`
- 新增配置项: `EXCEPTION_LOG_CODE` 触发记录文件日志的响应码, 默认 `[500]`
- 新增配置项: `SYS_ROLE_ID` 保留角色(权限组) ID 设置, 禁止将其他用户授权为该角色, 默认 `0`
- 用户表使用 `id` 字段配合 `flask_login` (原使用 `job_number` 字段)
- 用户权限组关联使用 `role_id` 字段 (原使用 `role` 标识)
- 日志表增加 `log_content` 字段, 用于记录日志详情, 日志记录方式见 ASNCharge.add()
- 数据库操作使用 `replace` 方法新增数据时, 确保总能获取到最新插入数据后的自增主键 ID
- 自定义信号机制(事件订阅-发布), 使用见示例代码: 系统管理事件, 用户登录成功事件
- 封装了邮件发送类, 基于 `flask_mail`, 支持邮件内容按模板渲染(HTML, TEXT), 支持异步发送邮件(默认)
- 附带了一个通用的邮件模板, 通过手机系统邮件/钉钉/QQ/微信/Foxmail/Outlook等邮件客户端兼容性测试



app/__init__.py
"""
    __init__.py
    ~~~~~~~~
    初始化 app
    注册蓝图到 app
    日志及环境配置

    :author: Fufu, 2019/9/2
"""
import logging
import os
from collections import namedtuple
from datetime import datetime, date
from decimal import Decimal

from concurrent_log_handler import ConcurrentRotatingFileHandler
from flask import Flask as _Flask, render_template
from flask.json import JSONEncoder as _JSONEncoder, jsonify
from werkzeug.middleware.proxy_fix import ProxyFix

from .forms import csrf
from .libs.exceptions import APIException, APIServerError, APIFailure
from .libs.helper import is_accept_json, get_int
from .models import db, DBModel
from .models.dbcache import init_db_cache
from .services.auth import oauth, login_manager
from .services.mail import mail
from .services.observers import observer
from .views import init_blueprint, init_template_global


class JSONEncoder(_JSONEncoder):
    """重写 Flask JSON encoder"""

    def default(self, o):
        """
        支持 DBModel 和日期时间. (注: ``UUID``, ``dataclasses`` 等未继承)

        e.g.::

            # {job_number: 114, mobile: "13880000009", last_login: "2019-09-10 09:04:59", status: 1}
            return jsonify(TBUser.query.get(114).hide_keys('realname', 'role_id'))

        :param o:
        :return:
        """
        if isinstance(o, datetime):
            return o.strftime('%Y-%m-%d %H:%M:%S')
        if isinstance(o, date):
            return o.strftime('%Y-%m-%d')
        if isinstance(o, Decimal):
            return float(o)
        if hasattr(o, 'keys') and hasattr(o, '__getitem__'):
            return dict(o)
        raise APIServerError()


class Flask(_Flask):
    json_encoder = JSONEncoder

    def make_response(self, rv):
        """
        支持视图函数直接返回 DBModel, list, dict(1.1.0+官方支持)

        e.g.::

            # {job_number: 114, mobile: "13880000009", last_login: "2019-09-10 09:04:59", status: 1}
            return TBUser.query.get(114).hide_keys('realname', 'role')

        :param rv:
        :return:
        """
        if isinstance(rv, set):
            rv = list(rv)
        if isinstance(rv, (list, dict, DBModel)):
            rv = jsonify(rv)
        return super(Flask, self).make_response(rv=rv)


def init_config(app, config_name=None, config=None):
    """
    加载各类环境配置

    :param app: Flask
    :param config_name: str, 指定配置文件名或环境变量配置名 e.g. development / testing / production
    :param config: str / dict 最高优先级配置, 字典或文件名
    :return: None
    """
    # 加载生产环境配置
    app.config.from_object('app.conf.secret_settings')
    app.config.from_object('app.conf.settings')

    # 加载指定名称的文件配置
    if not config_name:
        config_name = os.getenv(app.config.get('APP_NAME', 'FLASK_CONFIG'))
    if config_name:
        try:
            app.config.from_object('app.conf.{}_settings'.format(config_name))
        except ImportError:
            pass

    # 加载指定的参数配置
    if config is not None:
        if isinstance(config, dict):
            app.config.update(config)
        elif config.endswith('.py'):
            app.config.from_pyfile(config, silent=True)


def init_logger(app):
    """
    初始化日志(按大小滚动)

    :param app: Flask
    :return: None
    """
    log_maxsize = app.config.get('LOG_MAXSIZE', 100)
    log_backup = app.config.get('LOG_BACKUP', 20)
    log_level = app.config.get('LOG_LEVEL', logging.INFO)
    app_log = app.config.get('LOG_FILE')
    if not app_log:
        app_log = os.path.join(os.path.dirname(app.root_path), 'logs', 'app.log')
    fh = ConcurrentRotatingFileHandler(app_log, maxBytes=log_maxsize * 1024 * 1024, backupCount=log_backup,
                                       encoding='utf-8', use_gzip=True)
    fh.setLevel(log_level)
    fh.setFormatter(logging.Formatter(
        app.config.get('LOG_FORMAT', '%(asctime)s %(levelname)s %(module)s.%(funcName)s: %(message)s')))
    app.logger.addHandler(fh)
    app.logger.setLevel(logging.DEBUG if app.debug else log_level)


def init_error(app):
    """
    全局异常处理

    :param app: Flask
    :return: None
    """

    @app.errorhandler(Exception)
    def global_exception_handler(e):
        # 500 错误时记录日志
        code = getattr(e, 'code', 500)
        code in app.config.get('EXCEPTION_LOG_CODE', [500]) and app.logger.error(e, exc_info=True)

        if isinstance(e, APIException):
            return e
        if app.debug:
            raise e

        # 自定义异常描述
        err = namedtuple('MyError', ['code', 'description'])(
            code, getattr(e, 'description', app.config.get('EXCEPTION_DESC', {}).get(code, str(e)))
        )

        if app.config.get('API') or is_accept_json():
            return APIFailure(description=err.description)

        return render_template('base-msg.html', e=err), code


def create_app(config_name=None, config=None):
    """
    初始化 Flask 应用

    :param config_name: str
    :param config: str / dict 扩展配置或配置文件
    :return: Flask
    """
    app = Flask(__name__)

    # 加载环境配置
    init_config(app, config_name=config_name, config=config)

    x_for = get_int(app.config.get('PROXY_FIX_X_FOR'))
    if x_for:
        app.wsgi_app = ProxyFix(app.wsgi_app, x_for=x_for, x_host=x_for)

    # 日志配置
    init_logger(app)

    # 自定义错误页
    init_error(app)

    # 注册蓝图
    init_blueprint(app)

    # 注册 OAuth2
    oauth.init_app(app)

    # 注册 flask_login
    login_manager.init_app(app)

    # 注册邮件
    mail.init_app(app)

    # 注册自定义信号
    observer.init_app(app)

    # CSRF 保护
    csrf.init_app(app)

    # 注册模板全局函数
    init_template_global(app)

    # 初始化数据库配置
    db.init_app(app)
    with app.app_context():
        db.create_all()
        init_db_cache(app)

    @app.after_request
    def after_request(resp):
        if app.debug and app.config.get('DEBUG_RESPONSE') and is_accept_json():
            app.logger.debug("Response: {}, {}, {}".format(resp.status, resp.mimetype, resp.get_data(as_text=True)))
        return resp

    return app
